enh(adapater): add bigquery support

mmngreco opened the pull request:
I'm trying my best to include bigquery support for dadbod. I like to get some feedback before continue as I know there are some flaws (I'm not an vim-plugin expert). 

# In summary:

- Delegates into `bq` command (that has to be installed and configurated)
- `bq` has global and "local" options that can be set by the url e.g: `DB bigquery://?use_legacy_sql=True&dataset_id=publicdata:samples select * from shakespeare` (`dataset_id` is global, `use_legacy_sql` is local as only valid under `bq query` subcommand.

# `bq` Examples

- `bq --dataset_id=publicdata:samples query "select * from shakespeare limit 5" `
- `bq query "select * from publicdata.samples.shakespeare limit 5"`
- `bq ls` # list datasets

related #95 #104 #79

tpope commented on autoload/db/adapter/bigquery.vim:
```diff
@@ -0,0 +1,81 @@
+
+function! db#adapter#bigquery#input(url, in) abort
+
+  " QST: what is the purpose of input?
```
The purpose is to give the filename directly to the database, so that error messages, etc, can reference. Five adapters use this; the others accept input on stdin. Does bq really support neither?

mmngreco replied:
yes, it support it but I've faced some issue when I've tried to use it. I will review it.

tpope commented on autoload/db/adapter/bigquery.vim:
```diff
@@ -0,0 +1,81 @@
+
+  endfor
+
+  return join(cmd + subcmd)
```
You're building a string here, only to `split()` it back to an list in `#interactive()`. You should build a list to begin with.

tpope commented on autoload/db/adapter/bigquery.vim:
```diff
@@ -0,0 +1,81 @@
+endfunction
+
+function! db#adapter#bigquery#complete_opaque(url) abort
+  " QST: don't know why this is required
```
"Opaque" describes URLs like `mailto:user@example.com`, which do not adhere to the standard `schema://` pattern. Most Dadbod adapters accept both `adapter:dbname` and `adapter:///dbname`, which is why you'll commonly find one method delegating to the other.

mmngreco replied:
I see, then I can drop it, can't I?

tpope commented on autoload/db/adapter/bigquery.vim:
```diff
@@ -0,0 +1,81 @@
+if !executable('bq')
+  echom 'bq command is not available.'
+  let g:autoloaded_db_bigquery = 0
+endif
```
No other adapter does this; drop it. If you're unhappy with the experience when the db command is missing, that should be fixed in a more general way that works for all adapters.

mmngreco replied:
Fair point!

mmngreco commented:
@tpope let me know if it's ok for you, I think it's done.

mmngreco commented:
ping @tpope

tpope commented:
* `g:autoloaded_db_bigquery` still needs to be removed.
* Please use a `shiftwidth` of 2, like all other adapters. And use the built-in formatter, which does triple indenting for line continuations.
* I still see commented out code, a `FIXME`, and `echom` commands for debugging, which need to be addressed before merging.

mmngreco commented:
ups... sorry for that, it should be fixed now ;-)

tpope commented:
You fixed about half of what I said.

mmngreco commented:
I didn't read carefully your previous comment sorry for making you waste some time and energy on this. I've removed all comments and echoms.

I've also replaced the way a file is read as the bq parameter is not as robust as  I expected for doing this.

tpope commented:
> I've also replaced the way a file is read as the bq parameter is not as robust as I expected for doing this.

Sorry, but I'm not going to take ownership of a hack like that. Either it needs to read from stdin or a file, or it's not becoming an official adapter.

tpope commented:
Do note that Dadbod supports loading unofficial adapters from the runtime path.

mmngreco commented:
Ok, I think the issue was on my tests. I struggled a little when I had to debug what's going on... I will revert if it works as expected I will push if not I will convert it into unofficial adapater.

mmngreco commented:
I tried my best, after this commit you feel that this is not the place, feel free to close de PR. thank you for your time.

tpope commented on autoload/db/adapter/bigquery.vim:
```diff
@@ -0,0 +1,76 @@
+endfunction
+
+function! db#adapter#bigquery#input(url, in) abort
+  let out = []
+  if len(matchstr(a:in, '.sql'))
+    let out += db#adapter#bigquery#interactive(a:url, 'query')
+    let out += ['--flagfile='.a:in]
+  endif
+  return out
```
What is the purpose of the `len(matchstr(a:in, '.sql'))` check? If it fails, the whole function will return `[]`, which will break things.

mmngreco replied:
## Before that code:

![image](https://user-images.githubusercontent.com/6231413/205487665-a50aac27-362a-49ef-8713-8bf77d763e4a.png)

as `bq` fails if there is no query.

## After

![image](https://user-images.githubusercontent.com/6231413/205487534-49fbbb9f-aa09-417f-8cd9-29f0b9161486.png)

What I see is at the begining `db#adapter#bigquery#input` is called with the url and a file called `0` and the actual query use a file named `1.sql` . I don't know the better way to deal with this so I decided to use only files with the extension `.sql`.

and after that the adapter works as expected, what I've checked:

```
:DB bigquery://?project_id=siad-1141&format=sparse&use_legacy_sql=false select 1
:%DB bigquery://?project_id=siad-1141&format=sparse&use_legacy_sql=false 
:'<,'>DB bigquery://?project_id=siad-1141&format=sparse&use_legacy_sql=false 
```

tpope replied:
That `0` file is it doing an auth check. What happens if you pass it to `--flagfile` like an SQL file?

mmngreco replied:
it continues running endlessly.

```
$ touch 0
$  bq --project_id='siad-1141' query --format=sparse --use_legacy_sql=false --flagfile=./0
```

![image](https://user-images.githubusercontent.com/6231413/205498264-0de0b66b-ec52-42cc-97ec-612bafaf3361.png)

mmngreco replied:
> Is that because of the file name, or is it because the file is empty? 
it's because it's empty. 

> if you close stdin with < /dev/null, does it still hapen?

here's the output 

![image](https://user-images.githubusercontent.com/6231413/205507064-c3c3da0b-9fa0-441c-80c5-5ae5a27813f2.png)

tpope replied:
Remember this?

> The purpose is to give the filename directly to the database, so that error messages, etc, can reference. Five adapters use this; the others accept input on stdin. Does bq really support neither?

And this?

> Either it needs to read from stdin or a file, or it's not becoming an official adapter.

I have repeatedly brought up sending the query on stdin. And now we have extremely solid evidence that if you don't give it a query, it tries to read one from stdin.

Please show me a version without `db#adapter#bigquery#input()`. It's clearly causing more problems than it solves.

mbhynes commented:
Hi @mmngreco @tpope, thanks very much for your efforts on this so far!

I went through the review notes and the codebase, and I've opened up another take on this here: https://github.com/tpope/vim-dadbod/pull/125

I made several edits to the function interface (as well as fixing the `stdin` issues) such that this adapter could be used downstream by [`vim-dadbod-ui`](https://github.com/kristijanhusak/vim-dadbod-ui)

Hoping you folks might be able to review 🙏
